inc/Ui/Revisions.php

```diff
@@ -0,0 +1,330 @@
+
+        if (count($revisions) == 0 && $first != 0) {
+            $first = 0;
+            $revisions = $changelog->getRevisions($first, $conf['recent'] +1);
+        }
+        $hasNext = false;
+        if (count($revisions) > $conf['recent']) {
+            $hasNext = true;
+            array_pop($revisions); // remove extra log entry
+        }
+
+        // print intro
+        if (!$media_id) {
+            print p_locale_xhtml('revisions');
+            $exists = $INFO['exists'];
+            $display_name = useHeading('navigation') ? hsc(p_get_first_heading($id)) : $id;
+            if (!$display_name) {
+                $display_name = $id;
+            }
+        } else {
+            $exists = file_exists(mediaFN($id));
+            $display_name = $id;
+        }
+
+        // create the form
+        $form = new Form(['id' => 'page__revisions']);
+        $form->addClass('changes');
+        if ($media_id) {
+            $form->attr('action', media_managerURL(array('image' => $media_id), '&'));
+        }
+        $form->addTagOpen('div')->addClass('no');
+
+        // start listing
+        $form->addTagOpen('ul');
+
+        if ($exists && $first == 0) {
+            $minor = false;
+            if ($media_id) {
+                $date = dformat(@filemtime(mediaFN($id)));
+                $href = media_managerURL(array('image' => $id, 'tab_details' => 'view'), '&');
+
+                $changelog->setChunkSize(1024);
+                $revinfo = $changelog->getRevisionInfo(@filemtime(fullpath(mediaFN($id))));
+
+                $summary = $revinfo['sum'];
+                $editor = $revinfo['user'] ?: $revinfo['ip'];
+                $sizechange = $revinfo['sizechange'];
+            } else {
+                $date = dformat($INFO['lastmod']);
+                $sizechange = null;
+                if (isset($INFO['meta']) && isset($INFO['meta']['last_change'])) {
+                    if ($INFO['meta']['last_change']['type'] === DOKU_CHANGE_TYPE_MINOR_EDIT) {
+                        $minor = true;
+                    }
+                    if (isset($INFO['meta']['last_change']['sizechange'])) {
+                        $sizechange = $INFO['meta']['last_change']['sizechange'];
+                    }
+                }
+                $pagelog = new PageChangeLog($ID);
+                $latestrev = $pagelog->getRevisions(-1, 1);
+                $latestrev = array_pop($latestrev);
+                $href = wl($id, "rev=$latestrev", false, '&');
+                $summary = $INFO['sum'];
+                $editor = $INFO['editor'];
+            }
+
+            $form->addTagOpen('li')->addClass($minor ? 'minor' : '');
+            $form->addTagOpen('div')->addClass('li');
+            $form->addCheckbox('rev2[]')->val('current');
+            $form->addHTML(' ');
+
+            $form->addTagOpen('span')->addClass('data');
+            $form->addHTML($date);
+            $form->addTagClose('span');
+            $form->addHTML(' ');
```

Comment:
I had just know how to use and what code looks if Form\Form class used. I felt it might be convenient if I could write it in one line instead of 3 line, like:
``` php
            $form->addTagHTML('span')->addClass('date')->content($date);
```
The above idea may be implemented as, for example, Form\GenericElement class.
But, I also agree with you that it's not necessary to build whole form html using Form\Form class. In 67ef3e8 the revision list form will be generated with minimal use Form methods.

Reply:
I think a good question to ask is "would a plugin author want to address this specific element of the form? or would they want to insert something between these two elements?" if the answer is yes, then it makes sense to put those HTML nodes into their own form elements. If it is one logical unit, there is no need to split it into multiple form elements.

Reply:
> I think a good question to ask is "would a plugin author want to address this specific element of the form? or would they want to insert something between these two elements?"

My answer is "no". The Revision and Recent lists should be one logical unit. Even if plugin authors want to remove specific revision item from the list , it is not matter of ui and should be solved in other event handler.